### PR TITLE
feat: sort memory segment

### LINKED ISSUE #1210: feat: sort memory segment
## Why

The state being squashed, the state diff are sorted in descending order.
We need to make sure the trie diffs are also sorted in descending order as our commitment strategy assumes sorted keys (in the same way).

## What
We can use non-det for sorting an array.

## How
1/ Load all segment values in a Python object.
2/ Sort on a predicate.
3/ Load in memory sorted values
4/ Iterate on sorted values (with non-det, obtain index of item in initial segment)
5/ Verify equality 
6/ Verify iteration counter == initial length

### PR DESCRIPTION
**Pull request by Eikix**
closes #1210

### CODE REVIEW

File: cairo/ethereum/crypto/hash.cairo [hunk 1]
```diff
@@ -23,3 +24,16 @@ func keccak256{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: Keccak
     tempvar hash = Hash32(value=value);
     return hash;
 }
+
+func Hash32__eq__(a: Hash32, b: Hash32) -> bool {
```
Comment (enitrat) [inline cairo/ethereum/crypto/hash.cairo:28]: just call `Bytes32__eq__` it's the same type

File: python/mpt/src/mpt/utils.cairo [hunk 2]
```diff
@@ -179,3 +189,103 @@ func check_branch_node(node: BranchNode) {
     }
     return ();
 }
+
+func sort_AccountDiff(diff: AccountDiff) -> AccountDiff {
+    alloc_locals;
+    if (cast(diff.value, felt) == 0) {
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:195]: is it optional?

File: python/mpt/src/mpt/utils.cairo [hunk 3]
```diff
@@ -179,3 +189,103 @@ func check_branch_node(node: BranchNode) {
...
+        )
+    %}
+
+    tempvar sorted_diff = new AccountDiffStruct(
+        data=cast(buffer, AddressAccountNodeDiffEntry*), len=diffs_len
+    );
+    // 👆 Unknown ap change so we reassign it
+    tempvar sorted_diff = sorted_diff;
+    tempvar i = 0;
+    tempvar sorted_indexes = sorted_indexes;
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:230]: make it a local if it's not mutated in the loop

File: python/mpt/src/mpt/utils.cairo [hunk 4]
```diff
@@ -179,3 +189,103 @@ func check_branch_node(node: BranchNode) {
...
+            ids.sorted_indexes,
+            [int(item) for item in sorted_indexes]
+        )
+    %}
+
+    tempvar sorted_diff = new AccountDiffStruct(
+        data=cast(buffer, AddressAccountNodeDiffEntry*), len=diffs_len
+    );
+    // 👆 Unknown ap change so we reassign it
+    tempvar sorted_diff = sorted_diff;
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:228]: make it a local if not mutated in the loop

File: python/mpt/src/mpt/utils.cairo [hunk 5]
```diff
@@ -179,3 +189,103 @@ func check_branch_node(node: BranchNode) {
...
+
+    with_attr error_message("ValueError") {
+        let sorted_index = [sorted_indexes + i];
+
+        let prev_key = prev_diff.data[sorted_index].value.key.value;
+        let new_key = sorted_diff.data[i].value.key.value;
+        assert prev_key = new_key;
+
+        let prev_eq = AccountNode__eq__(
+            prev_diff.data[sorted_index].value.prev_value, sorted_diff.data[i].value.prev_value
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:246]: you should be able to compare pointers instead?

**Comment by Eikix** [comment]
Next steps:
1/ pass test
2/ implement hints

File: cairo/ethereum/crypto/hash.cairo (hunk 1 above)
Comment (Eikix) [reply cairo/ethereum/crypto/hash.cairo:28]: didn't we want to differentiate between types?

Comment (enitrat) [reply cairo/ethereum/crypto/hash.cairo:28]: A hash32 is an alias to a bytes32

File: python/mpt/src/mpt/utils.cairo [hunk 6]
```diff
@@ -179,3 +191,90 @@ func check_branch_node(node: BranchNode) {
...
+        )
+    %}
+
+    tempvar sorted_account_diffs = new AccountDiffStruct(
+        data=cast(buffer, AddressAccountNodeDiffEntry*), len=diffs_len
+    );
+    tempvar i = 0;
+    local sorted_account_diffs: AccountDiffStruct* = sorted_account_diffs;
+    local sorted_indexes: felt* = sorted_indexes;
+    local range_check_ptr = range_check_ptr;
```
[the remark this answers is not in the document]
Comment (Eikix) [reply python/mpt/src/mpt/utils.cairo:229]: oops.. just saw this (spoiler alert i got issues)

File: python/mpt/src/mpt/utils.cairo [hunk 7]
```diff
@@ -179,3 +191,90 @@ func check_branch_node(node: BranchNode) {
...
+    with_attr error_message("ValueError") {
+        if (is_end == 0) {
+            // If we are not at the end,
+            // We can access offset i + 1
+            let is_ordered = is_le_felt(
+                sorted_account_diffs.data[i].value.key.value,
+                sorted_account_diffs.data[i + 1].value.key.value,
+            );
+            assert is_ordered = 1;
+            tempvar range_check_ptr = range_check_ptr;
```
[the remark this answers is not in the document]
Comment (Eikix) [reply python/mpt/src/mpt/utils.cairo:262]: > assert [range_check_ptr] = sorted_account_diffs.data[i + 1].value.key.value - (sorted_account_diffs.data[i].value.key.value + 1)

why?

[the remark this answers is not in the document]
Comment (enitrat) [reply python/mpt/src/mpt/utils.cairo]: simpler

[the remark this answers is not in the document]
Comment (enitrat) [reply python/mpt/src/mpt/utils.cairo]: ah wait nevermind, the key is not in a u128

File: python/mpt/src/mpt/utils.cairo [hunk 8]
```diff
@@ -179,3 +190,77 @@ func check_branch_node(node: BranchNode) {
...
+    jmp end if is_end != 0;
+
+    // Check that the sorted array is ordered
+    with_attr error_message("ValueError") {
+        // If we are not at the end,
+        // We can access offset i + 1
+        let is_descending_ordered = is_le_felt(
+            sorted_account_diffs.data[i].value.key.value,
+            sorted_account_diffs.data[i - 1].value.key.value,
+        );
```
[the remark this answers is not in the document]
Comment (enitrat) [reply python/mpt/src/mpt/utils.cairo:248]: assuming that of course there are not duplicates

File: python/mpt/src/mpt/utils.cairo [hunk 9]
```diff
@@ -179,3 +190,77 @@ func check_branch_node(node: BranchNode) {
...
+    tempvar i = i + 1;
+    jmp end if is_end != 0;
+
+    // Check that the sorted array is ordered
+    with_attr error_message("ValueError") {
+        // If we are not at the end,
+        // We can access offset i + 1
+        let is_descending_ordered = is_le_felt(
+            sorted_account_diffs.data[i].value.key.value,
+            sorted_account_diffs.data[i - 1].value.key.value,
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:247]: you incremented `i` right before. so this is `i, i+1` ? misleading to shadow the `i` variable I think

File: python/mpt/src/mpt/utils.cairo [hunk 10]
```diff
@@ -179,3 +190,77 @@ func check_branch_node(node: BranchNode) {
...
+    end:
+    // Check that the loop has been executed the correct number of times
+    // Such that we know len(sorted) == len(initial_array)
+    let i = [ap - 1];
+    assert i = diffs_len;
+
+    let res = AccountDiff(sorted_account_diffs);
+
+    return res;
+}
```
Comment (enitrat) [inline python/mpt/src/mpt/utils.cairo:266]: my full proposal:

```suggestion
// @notice Sorts an AccountDiff struct in descending order based on the key.
// This function implies that the original AccountDiff struct does not contain any duplicate keys.
// @dev The sorted segment is returned from the hint.
// Verifications:
// - The sorted segment is in strict descending order based on the key.
//    (forall i in [0, diffs_len - 1], sorted_diff_struct.data[i].value.key > sorted_diff_struct.data[i + 1].value.key)
// - The sorted segment is a permutation of the original segment (
//    (forall i in [0, diffs_len - 1], exists j in [0, diffs_len - 1] such that sorted_diff_struct.data[i].value.key = original_diff_struct.data[j].value.key)
//    AND len(sorted_diff_struct) = len(original_diff_struct)
// @param account_diff The AccountDiff struct to sort.
// @return The sorted AccountDiff struct.
func sort_AccountDiff{range_check_ptr}(account_diff: AccountDiff) -> AccountDiff {
    alloc_locals;
    let diffs_len = account_diff.value.len;
    if (diffs_len == 0) {
        // If the input diff is empty, it's already sorted
        return account_diff;
    }

    // Pointer to the original, unsorted data
    tempvar diffs_ptr = account_diff.value.data;

    // A map(sorted_index -> original_index) to store the original index corresponding to each entry in the sorted buffer
    let (buffer) = alloc();
    let (local sorted_to_original_index_map) = alloc();

    %{ sort_account_diff %}

    tempvar sorted_diff_struct = new AccountDiffStruct(
        data=cast(buffer, AddressAccountNodeDiffEntry*), len=diffs_len
    );
    local sorted_diff_struct_ptr: AccountDiffStruct* = sorted_diff_struct;

    tempvar range_check_ptr = range_check_ptr;
    tempvar loop_counter = 0;


    loop:
    let range_check_ptr = [ap - 2];
    let loop_counter = [ap - 1];
    let original_diff_struct_ptr = cast([fp - 3], AccountDiffStruct*);

    // --- Verification Step 1: Permutation Check ---
    // Ensure that the element at the current sorted position (`loop_counter`)
    // corresponds exactly to an element from the original array, using the
    // `original_index` provided by the hint's `original_index_map`.
    with_attr error_message("ValueError: Sorted element does not match original element at hint index") {
        let original_index = [sorted_to_original_index_map + loop_counter];

        tempvar original_entry: AddressAccountNodeDiffEntry = original_diff_struct_ptr.data[original_index];
        tempvar sorted_entry: AddressAccountNodeDiffEntry = sorted_diff_struct_ptr.data[loop_counter];

        // Identical keys & struct pointers at this index
        assert original_entry.value.key.value = sorted_entry.value.key.value;
        assert sorted_entry.value = original_entry.value;
    }

    // `diffs_len - loop_counter - 1` is safe because diffs_len >= 1 and loop_counter starts at 0.
    let is_last_element = is_zero(original_diff_struct_ptr.len - loop_counter - 1);

    tempvar next_loop_counter = loop_counter + 1;
    jmp end if is_last_element != 0;

    // --- Verification Step 2: Ordering Check ---
    // Ensure that the sorted array is in strict descending order based on the key.
    // This check is performed for elements from index 1 up to diffs_len - 1.
    with_attr error_message("ValueError: Array is not sorted in descending order") {
        let next_key = sorted_diff_struct_ptr.data[next_loop_counter].value.key.value;
        let previous_key = sorted_diff_struct_ptr.data[loop_counter].value.key.value;
        let keys_ordered = is_le_felt(next_key, previous_key);
        let keys_not_equal = is_not_zero(next_key - previous_key);

        assert keys_ordered * keys_not_equal = 1;
    }

    tempvar range_check_ptr = range_check_ptr;
    tempvar loop_counter = next_loop_counter;
    jmp loop;

    end:
    let final_loop_counter = [ap - 1];
    // --- Verification Step 3: Loop Count Check ---
    // Ensure that the loop executed exactly `diffs_len` times, confirming that
    // all elements were processed and the length of the sorted array matches the original.
    assert final_loop_counter = diffs_len;

    let sorted_account_diff = AccountDiff(sorted_diff_struct_ptr);

    return sorted_account_diff;
}

```

renamed sorted_indexes to `sorted_to_original_index_map` - to adapt in the hint

File: python/mpt/tests/src/mpt/test_utils.py [hunk 11]
```diff
@@ -74,3 +77,66 @@ def test_check_branch_node_invalid_x_branch(
...
+        self, cairo_run, data: List[AddressAccountNodeDiffEntry]
+    ):
+        sorted_data = sorted(
+            data, key=lambda x: int.from_bytes(x.key, "little"), reverse=True
+        )
+        cairo_data = cairo_run("sort_account_diff", data)
+        assert cairo_data == sorted_data
+
+    @given(data=...)
+    def test_sort_account_diff_ascending_order(
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:99]:
```suggestion
    def test_sort_account_diff_should_fail_if_not_descending_order(
```

File: python/mpt/tests/src/mpt/test_utils.py [hunk 12]
```diff
@@ -74,3 +77,66 @@ def test_check_branch_node_invalid_x_branch(
...
+
+# Sort pointers based on the key values they point to, in ascending order
+sorted_pointers = sorted(pointers, key=lambda ptr: memory[ptr])
+
+# Load the sorted pointers into ids.buffer
+segments.load_data(ids.buffer, sorted_pointers)
+
+indices = list(range(ids.diffs_len))
+sorted_indices = sorted(indices, key=lambda i: memory[pointers[i]], reverse=True)
+segments.load_data(ids.sorted_indexes, sorted_indices)
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:117]:
```suggestion
    pointers = [memory[ids.diffs_ptr.address_ + i] for i in range(ids.diffs_len)]
    sorted_pointers = sorted(pointers, key=lambda ptr: memory[ptr], reverse=True)
    # Invert the order of the last two elements
    sorted_pointers[-2], sorted_pointers[-1] = sorted_pointers[-1], sorted_pointers[-2]
    segments.load_data(ids.buffer, sorted_pointers)
    indices = list(range(ids.diffs_len))
    sorted_indices = sorted(indices, key=lambda i: memory[pointers[i]], reverse=True)
    segments.load_data(ids.sorted_to_original_index_map, sorted_indices)
```

make the change explicit on a slight variation of the input

File: python/mpt/tests/src/mpt/test_utils.py [hunk 13]
```diff
@@ -74,3 +77,66 @@ def test_check_branch_node_invalid_x_branch(
...
+
+# Load the sorted pointers into ids.buffer
+# Repeat the first element n times to create a different list
+first_element = sorted_pointers[0]
+repeated_list = [first_element] * ids.diffs_len
+segments.load_data(ids.buffer, repeated_list)
+
+indices = list(range(ids.diffs_len))
+sorted_indices = sorted(indices, key=lambda i: memory[pointers[i]], reverse=True)
+segments.load_data(ids.sorted_indexes, sorted_indices)
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:145]:
```suggestion
    pointers = [memory[ids.diffs_ptr.address_ + i] for i in range(ids.diffs_len)]
    sorted_pointers = sorted(pointers, key=lambda ptr: memory[ptr], reverse=True)
    # BAD HINT: not a permutation of the input list
    sorted_pointers[-1] += 1
    segments.load_data(ids.buffer, sorted_pointers)
    indices = list(range(ids.diffs_len))
    sorted_indices = sorted(indices, key=lambda i: memory[pointers[i]], reverse=True)
    segments.load_data(ids.sorted_to_original_index_map, sorted_indices)
```

Make the change targeted and explicit

File: python/mpt/tests/src/mpt/test_utils.py [hunk 14]
```diff
@@ -74,3 +77,66 @@ def test_check_branch_node_invalid_x_branch(
...
+            """,
+        ):
+            with strict_raises(KeyError):
+                cairo_run_py("sort_account_diff", data)
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:149]: also add a test where the input list is a list that is already sorted but contains duplicate values: it should also fail.

File: python/mpt/tests/src/mpt/test_utils.py [hunk 15]
```diff
@@ -15,6 +15,13 @@
 from mpt.utils import check_branch_node, nibble_list_to_bytes
 from tests.utils.args_gen import AddressAccountNodeDiffEntry
 
+list_address_account_node_diff_entry_strategy = st.lists(
+    st.from_type(AddressAccountNodeDiffEntry),
+    min_size=1,
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:20]:
```suggestion
    min_size=0,
```

any reason to use a non-empty list?

Comment (Eikix) [reply python/mpt/tests/src/mpt/test_utils.py]: the patched hints will not raise

Comment (enitrat) [reply python/mpt/tests/src/mpt/test_utils.py]: then you should not be using this strategy in the happy path test

**Comment by Eikix** [comment]
Thanks for the detailed suggestions. Way better now

**Comment by enitrat** [comment]
@Eikix can you address the suggestions on the strategies to use and the patched hints?

File: python/mpt/tests/src/mpt/test_utils.py [hunk 16]
```diff
@@ -74,3 +84,122 @@ def test_check_branch_node_invalid_x_branch(
...
+                Exception,
+                match="ValueError: Sorted element does not match original element at hint index",
+            ):
+                cairo_run_py("sort_account_diff", data)
+
+    @given(data=list_address_account_node_diff_entry_strategy)
+    def test_sort_account_diff_sorted_list_with_duplicates(
+        self, cairo_programs, cairo_run_py, data: List[AddressAccountNodeDiffEntry]
+    ):
+        with patch_hint(
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:159]: no need for patch hint in this test. you need to change the strategy to have duplicates

File: python/mpt/tests/src/mpt/test_utils.py [hunk 17]
```diff
@@ -74,3 +84,122 @@ def test_check_branch_node_invalid_x_branch(
...
+
+    @given(data=list_address_account_node_diff_entry_strategy)
+    def test_sort_account_diff_should_fail_if_not_descending_order(
+        self, cairo_programs, cairo_run_py, data: List[AddressAccountNodeDiffEntry]
+    ):
+        with patch_hint(
+            cairo_programs,
+            "sort_account_diff",
+            """
+# Extract the list of pointers directly
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:106]: remove comments not related to the changes of the patched hint

File: python/mpt/tests/src/mpt/test_utils.py [hunk 18]
```diff
@@ -74,3 +84,122 @@ def test_check_branch_node_invalid_x_branch(
         ):
             with strict_raises(ValueError):
                 cairo_run_py("check_branch_node", branch_node)
+
+    @given(data=list_address_account_node_diff_entry_strategy)
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:88]: this should accept lists of any size, not [2, [

File: python/mpt/tests/src/mpt/test_utils.py [hunk 19]
```diff
@@ -74,3 +99,105 @@ def test_check_branch_node_invalid_x_branch(
...
+# BAD HINT: Invert the order of the last two elements
+sorted_pointers[-2], sorted_pointers[-1] = sorted_pointers[-1], sorted_pointers[-2]
+
+segments.load_data(ids.buffer, sorted_pointers)
+
+indices = list(range(ids.diffs_len))
+sorted_to_original_index_map = sorted(indices, key=lambda i: memory[pointers[i]], reverse=True)
+
+# BAD HINT: Invert the order of the last two elements
+sorted_to_original_index_map[-2], sorted_to_original_index_map[-1] = sorted_to_original_index_map[-1], sorted_to_original_index_map[-2]
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:133]: ah yes, nice one

File: python/mpt/tests/src/mpt/test_utils.py (hunk 16 above)
Comment (enitrat) [reply python/mpt/tests/src/mpt/test_utils.py:174]: up

File: python/mpt/tests/src/mpt/test_utils.py [hunk 20]
```diff
@@ -74,3 +99,106 @@ def test_check_branch_node_invalid_x_branch(
...
+
+    @given(data=list_address_account_node_diff_entry_strategy_with_duplicates())
+    def test_sort_account_diff_sorted_list_with_duplicates(
+        self, cairo_run_py, data: List[AddressAccountNodeDiffEntry]
+    ):
+
+        with strict_raises(
+            Exception, match="ValueError: Array is not sorted in descending order"
+        ):
+            cairo_run_py("sort_account_diff", data)
```
Comment (enitrat) [inline python/mpt/tests/src/mpt/test_utils.py:178]:
```suggestion
        self, cairo_run, data: List[AddressAccountNodeDiffEntry]
    ):

        with strict_raises(
            Exception, match="ValueError: Array is not sorted in descending order"
        ):
            cairo_run("sort_account_diff", data)
```

**Review by enitrat — APPROVED** [review]
